refactor(chain)!: Change tx_last_seen to `Option<u64>`

ValuedMammal opened the pull request:
The PR changes the type of last_seen to `Option<u64>` for `txs` member of `TxGraph`.

This fixes an issue where unbroadcast and otherwise non-canonical transactions were returned from methods `list_chain_txs` and `Wallet::transactions` because every new tx inserted had a last_seen of 0 making it appear unconfirmed.

fixes #1446 
fixes #1396 

### Notes to the reviewers

### Changelog notice

Changed
- Member `last_seen_unconfirmed` of `TxNode` is changed to `Option<u64>`
- Renamed `TxGraph` method `list_chain_txs` to `list_canonical_txs`
- Changed `Wallet::insert_tx` to take a single `tx: Transaction` as parameter

Added
- Add method `txs_with_no_anchor_or_last_seen` for `TxGraph`
- Add method `unbroadcast_transactions` for `Wallet`

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### Bugfixes:

* [x] This pull request breaks the existing API
* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

ValuedMammal commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -959,10 +959,10 @@ fn test_chain_spends() {
     );
 
-    // Even if unconfirmed tx has a last_seen of 0, it can still be part of a chain spend.
+    // Even if unconfirmed tx has a last_seen of None, it can still be part of a chain spend.
```
Is this a valid assumption?

If instead we were to filter unseen txs in `try_get_chain_position`, it has the same desired effect on canonical txs, but then this assertion no longer holds. In other words, the above comment will say
```
// An unconfirmed and unseen tx cannot be part of a chain spend.
```

Experimenting with that approach in 03d0515ec7a1712c1614d0b0e0947e97463e816d. This might be preferable in that it applies the same reasoning (that an unseen tx does not have a chain position) to `list_chain_spends` and `filter_chain_txouts`.

storopoli commented:
Isn't `Default` for `Option<T>` `None`? If yes, maybe we can simplify with `..Default::default()`?

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -794,7 +794,15 @@ impl<A: Anchor> TxGraph<A> {
+                    return Ok(None);
+                }
+                val
+            }
```
I like the idea of moving this logic into `try_get_chain_position`. However, for transactions with `last_seen` as `None`, but with an anchor that anchors the transaction in the best chain, wouldn't the transaction be in the best chain?

My understanding of `last_seen` is that a value of `None` signals that it has not been seen in the mempool. However, this fact shouldn't invalidate the fact that the transaction can be confirmed in the best chain?

evanlinjin replied:
Should we rename this method to `try_get_canonical_position`?

ValuedMammal replied:
I think having a chain position should imply the tx is canonical, so I think these are synonymous.

evanlinjin commented on crates/chain/src/chain_data.rs:
```diff
@@ -10,7 +10,7 @@ pub enum ChainPosition<A> {
     Confirmed(A),
     /// The chain data is not confirmed and last seen in the mempool at this timestamp.
-    Unconfirmed(u64),
+    Unconfirmed(Option<u64>),
```
Would it be fair to say that unconfirmed transactions without a last-seen will not have a `ChainPosition`? Would it make sense to rename `ChainPosition` to `CanonicalPosition`?

evanlinjin commented on crates/chain/src/chain_data.rs:
```diff
@@ -58,14 +58,16 @@ pub enum ConfirmationTime {
     Unconfirmed {
         /// The last-seen timestamp in unix seconds.
-        last_seen: u64,
+        last_seen: Option<u64>,
```
Maybe it's not worth it to put an option here either? Assuming that we agree that unconfirmed txs without a last-seen are not worth returning? The `ConfirmationTime` type is only used in `bdk_wallet` as a field in `LocalOutput` (which is a UTXO). There is no point showing a UTXO that is not part of the canonical chain right?

ValuedMammal replied:
> There is no point showing a UTXO that is not part of the canonical chain right?

I'm leaning towards no, a utxo owned by the wallet shouldn't qualify as a `LocalOutput` if it doesn't come from a tx in the canonical chain.

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -965,18 +978,24 @@ impl<A: Anchor> TxGraph<A> {
-    pub fn try_list_chain_txs<'a, C: ChainOracle + 'a>(
+    /// [`full_txs`]: Self::full_txs
+    /// [`canonical_transactions`]: Self::canonical_transactions
+    pub fn try_canonical_transactions<'a, C: ChainOracle + 'a>(
```
Would `try_list_canonical_transactions` be a better name?

evanlinjin commented on crates/chain/tests/common/tx_template.rs:
```diff
@@ -131,9 +131,7 @@ pub fn init_graph<'a, A: Anchor + Clone + 'a>(
-        if let Some(seen_at) = tx_tmp.last_seen {
-            let _ = graph.insert_seen_at(tx.txid(), seen_at);
-        }
+        let _ = graph.insert_seen_at(tx.txid(), tx_tmp.last_seen.unwrap_or(0));
```
Why are we force-inserting a `last_seen` when the template `last_seen` is `None`?

ValuedMammal replied:
This was a fix for `test_tx_graph_conflicts`. Because we're interested in testing tx conflicts, we have to control for the fact that txs without a last_seen would now be filtered out, so we just assume all txs are canonical.

edit: An alternative is to patch up some of the scenarios in `test_tx_graph_conflicts` to ensure every tx template is given some value of last_seen. That might make things clearer.

evanlinjin commented on crates/wallet/src/wallet/coin_selection.rs:
```diff
@@ -784,13 +784,21 @@ mod test {
 
     fn get_test_utxos() -> Vec<WeightedUtxo> {
```
These aren't really useful UTXOs when they are not seen at all right?

evanlinjin commented:
@LLFourn's [comment](https://github.com/bitcoindevkit/bdk/issues/1396#issuecomment-2036248375)...

> `None` would be in no ones mempool (not even yours). You can have a transaction that is valid but has never left application memory.

summarizes the main usecase of these changes.

With this context in mind, this PR description needs to highlight how we should handle non-broadcasted transactions in `TxGraph`. How should these non-broadcasted transactions influence methods `filter_chain_txouts`/`filter_chain_unspents`-esc methods.

How can the caller list these non-broadcasted transactions? What if the caller would like to include UTXOs of non-broadcasted transactions when listing their UTXO set? What if the caller would NOT like to do so? What if the caller would like to replace non-broadcasted transactions? Currently, we do transaction-precedence based on last_seen. Wouldn't we need something like this for non-broadcasted transactions too?

It seems that many things have not be thought through clearly. Let's have some more concrete discussion about how everything should look after these changes.

ValuedMammal committed:
fix(wallet)!: Rework `Wallet::insert_tx` to no longer insert anchors

since we'd be lacking context that should normally occur during
sync with a chain source. The logic for inserting a graph
anchor from a `ConfirmationTime` is moved to the wallet common
test module in order to simulate receiving new txs and
confirming them.

ValuedMammal committed:
fix(tx_graph)!: Change tx_last_seen to `Option<u64>`

Also fixup `test_list_owned_txouts` to check that the right
outputs, utxos, and balance are returned at different local
chain heights.

This fixes an issue where unbroadcast and otherwise non-canonical
transactions were returned from methods `list_chain_txs` and
`Wallet::transactions` because every tx inserted had a last_seen
of 0 making it appear unconfirmed.

Note this commit changes the way `Balance` is represented due to
new logic in `try_get_chain_position` that no longer considers
txs with non-canonical anchors. Before this change, a tx anchored
to a block that is reorged out had a permanent effect on the
pending balance, and now only txs with a last_seen time or an
anchor confirmed in the best chain will return a `ChainPosition`.

ValuedMammal committed:
test(wallet): Add test_insert_tx_balance_and_utxos

ValuedMammal committed:
test(indexed_tx_graph): Add test_get_chain_position

ValuedMammal committed:
ref(tx_graph)!: Rename `list_chain_txs` to `list_canonical_txs`

ValuedMammal commented:
While it's possible to store a transaction in `TxGraph` before broadcasting it, I hesitate to advertise this as a feature, because there's not an easy way to make any owned txouts available for coin selection. The design of `TxGraph` requires all txs have a canonical chain position in order to count toward the total balance. You can manipulate the graph into considering a transaction canonical, but without a more robust way to abandon a tx and restore an input's unspent status, the only way to revert it is to do a full wallet rescan.

As an aside, if you're trying to build a chain of dependent transactions without broadcasting them, it should be possible to bypass coin selection by adding any owned txouts as a foreign utxo to a new `TxBuilder`.

evanlinjin commented:
Another priority review for next week.

notmandatory commented:
@ValuedMammal if this is already rebased and ready to go please change the status to "ready to review".

evanlinjin commented:
> This fixes an issue where unbroadcast and otherwise non-canonical transactions were returned from methods 

How would a non-canonical tx be returned by the methods?

notmandatory commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -2299,6 +2228,16 @@ impl Wallet {
 
+    /// Get a mutable reference to the inner [`TxGraph`].
+    ///
+    /// **WARNING**: It's not recommended to make changes manually to the transaction
```
❓ Is this only being exposed for testing purposes? If so can we avoid move the tests that need it into a test module within this crate so it can be made `pub(crate)` instead? I hate to leave this foot-gun out there in out public API where is can (and likely will) be misused. 

Alternatively can existing "safe" APIs be used in tests instead of adding this function?

ValuedMammal replied:
I agree I don't like that it is `pub`. What do you think about having methods `insert_anchor` and `insert_seen_at` for Wallet? If not, we can try moving the tests that need it

notmandatory replied:
~~If this function is really only needed for tests can you leave it public but feature gate it with `#[cfg(test)]` ?Otherwise if `insert_anchor` and `insert_seen_at` can do their thing safely I like that approach, but again if they're only needed for testing I'd rather they not be in the public API.~~ 

Sorry for my slow reply, I see you already added these functions. They don't look any more dangerous than constructing a bogus wallet `Update` and applying it with `apply_update()`.  I still worry about complicating the wallet API, do you envision any other uses for these new functions except for testing? If not then how do you feel about putting them behind the `#[cfg(test)]` feature flag?

ValuedMammal replied:
I think that means moving a lot of things back to unit tests including everything that uses `get_funded_wallet`. Should I do this in a separate PR?

notmandatory replied:
~~Can't you leave the tests were they are and just hide those new functions from the public API with the `#[cfg(test)]` ? I haven't tried it but I think it should work that way.~~ Drat, I just learned that `#[cfg(test)]` functions are not available for the integration tests, I guess you were telling me that. I'm ok with leaving the functions in the public API for now then as long as they're reasonably safe to use.

evanlinjin replied:
I think it's okay to leave potentially-dangerous methods in wallet as long as it's documented. Potentially-dangerous also means powerful (sometimes).

ValuedMammal replied:
Resolved by removing `insert_anchor` and `insert_seen_at`, using `apply_update` instead.

notmandatory reviewed (commented):
Overall looks good to me, just one question.

ValuedMammal commented:
> > This fixes an issue where unbroadcast and otherwise non-canonical transactions were returned from methods
> 
> How would a non-canonical tx be returned by the methods?

For example in `test_list_owned_txouts` tx3 is registered as spending tx2 at block 1 when tx3 doesn't confirm until block 2 and is never explicitly given a last_seen in the test. Does that seem unexpected?

evanlinjin commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -906,6 +877,23 @@ impl Wallet {
+    pub fn insert_anchor(&mut self, txid: Txid, anchor: ConfirmationTimeHeightAnchor) {
+        let indexed_graph_changeset = self.indexed_graph.insert_anchor(txid, anchor);
+        self.stage.append(indexed_graph_changeset.into());
+    }
```
I feel like we should include a warning that the caller is responsible for ensuring the correctness of the information provided. If the caller anchors a tx in a block where the tx does not actually belong in, the canonical history of the wallet may be messed up permanently (since the `TxGraph` is monotone).

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -971,10 +981,10 @@ impl<A: Anchor> TxGraph<A> {
-    /// [`list_chain_txs`]: Self::list_chain_txs
-    pub fn try_list_chain_txs<'a, C: ChainOracle + 'a>(
+    /// [`list_canonical_transactions`]: Self::list_canonical_transactions
+    pub fn try_list_canonical_transactions<'a, C: ChainOracle + 'a>(
```
Many of our methods use `tx` as a name for `transaction`. What do you think about renaming this to `try_list_canonical_txs` for consistency?

evanlinjin commented:
I feel like we should include a method for returning "unseen and unanchored" transactions. Not sure what to name it, maybe `unseen_txs` is enough? I guess this would be on `TxGraph` and re-exposed on `Wallet`. The intention is to have a method that can be helpful for determining which txs still need to be broadcasted.

ValuedMammal committed:
wallet: delete method `insert_anchor`

ValuedMammal committed:
test(tx_graph): Add test for `list_canonical_txs`

ValuedMammal committed:
feat(tx_graph): Add method `txs_with_no_anchor_or_last_seen`

notmandatory reviewed (approved):
ACK 42fab2bcc6ba5f0df0385c017851d3589e7575cd

This looks good and all comments addressed.

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -110,7 +110,7 @@ use core::{
 pub struct TxGraph<A = ()> {
     // all transactions that the graph is aware of in format: `(tx_node, tx_anchors, tx_last_seen)`
-    txs: HashMap<Txid, (TxNodeInternal, BTreeSet<A>, u64)>,
+    txs: HashMap<Txid, (TxNodeInternal, BTreeSet<A>, Option<u64>)>,
```
This should not be in this tuple I think. `last_seen` should just been its own `HashMap<Txid, u64>` if the value would be `None` here it would just not have an entry in the map. This is more idomatic and easier to reason about imo.

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -258,6 +258,27 @@ impl<A> TxGraph<A> {
+        F: FnMut(TxNode<'a, Arc<Transaction>, A>) -> Option<O> + 'a,
+    {
+        self.full_txs().filter_map(f)
+    }
```
This merthod has no reason to exist. Anyone cal just call `full_txs()` and then `.filter_map` on the output.

LLFourn commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -1117,7 +1107,58 @@ fn update_last_seen_unconfirmed() {
+}
+
+#[test]
+fn list_canonical_txs() {
```
:pick: It would be good if this name could give us more information about what this is testing `transactions_inserted_into_tx_graph_are_not_canonical_until_they_have_an_anchor_in_best_chain`. I know we need to do better on this but we might as well start here.

LLFourn commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -2321,6 +2250,14 @@ impl Wallet {
+    pub fn unbroadcast_transactions(
+        &self,
+    ) -> impl Iterator<Item = TxNode<'_, Arc<Transaction>, ConfirmationTimeHeightAnchor>> {
+        self.as_ref().txs_with_no_anchor_or_last_seen()
```
`as_ref` is very ambiguous here

ValuedMammal replied:
using `tx_graph` method on Wallet instead

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -560,7 +581,7 @@ impl<A: Clone + Ord> TxGraph<A> {
         let mut update = Self::default();
         let (_, _, update_last_seen) = update.txs.entry(txid).or_default();
-        *update_last_seen = seen_at;
+        let _ = update_last_seen.insert(seen_at);
```
More idomatic just to assign to `Some(seen_at)`.

LLFourn reviewed (dismissed):
Thanks. I've made a few suggestions about how to clean up the internals and tests. API looks good.

ValuedMammal commented:
I'm interested in implementing the suggestion to make last-seen a new field in `TxGraph` that is `HashMap<Txid, u64>` if it can be done in this PR cleanly

ValuedMammal committed:
ref(tx_graph): Change last_seen to `HashMap<Txid, u64>`

notmandatory reviewed (approved):
Re ACK af75817d4bedbb5e148812d1073fd0729a23358b

@LLFourn (or anyone else) any other suggestions before we merge this one?
